### ISSUE TITLE
Kubelet container being constantly restarted due to "Env has changed from"

### ISSUE THREAD

**Issue by Oats87** [rancher/rancher#17502]
*What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
Haven't determined this yet

**Result:**
Kubelet is constantly being restarted

**Other details that may be helpful:**

`cattle-node-agent` logs are:
```
time="2019-01-21T20:32:34Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac] to [RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T20:38:40Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T20:40:42Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac] to [RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T20:46:48Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T20:52:54Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T20:54:56Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T21:01:03Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac] to [RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T21:03:04Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T21:05:06Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T21:11:13Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T21:13:15Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
time="2019-01-21T21:21:22Z" level=info msg="For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=d07839d401a8fb104bf77e6f336f0dd0] to [RKE_CLOUD_CONFIG_CHECKSUM=17e5175e20f80bd614d12c0601cce2ac PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]"
```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.1.4
- Installation option (single install/HA): HA

Kubernetes Version: "v1.11.5-rancher1-1"

Cluster.yml is:
```
addon_job_timeout: 30
authentication: 
  strategy: "x509"
bastion_host: 
  ssh_agent_auth: false
cloud_provider: 
  name: "vsphere"
  vsphereCloudProvider: 
    global: 
      datacenters: "<REDACTED>"
      insecure-flag: true
      password: "<REDACTED>"
      soap-roundtrip-count: 0
      user: "<REDACTED>"
    virtual_center: 
      <REDACTED>: 
        password: "<REDACTED>"
        soap-roundtrip-count: 0
        user: "<REDACTED>"
      <REDACTED>: 
        password: "<REDACTED>"
        soap-roundtrip-count: 0
        user: "<REDACTED>"
    workspace: 
      datacenter: "<REDACTED>"
      default-datastore: "<REDACTED>"
      folder: "<REDACTED>"
      resourcepool-path: "<REDACTED>"
      server: "<REDACTED>"
ignore_docker_version: false
# 
#   # Currently only nginx ingress provider is supported.
#   # To disable ingress controller, set `provider: none`
#   # To enable ingress on specific nodes, use the node_selector, eg:
#      provider: nginx
#      node_selector:
#        app: ingress
# 
ingress: 
  provider: "nginx"
kubernetes_version: "v1.11.5-rancher1-1"
monitoring: 
  provider: "metrics-server"
# 
#   # If you are using calico on AWS
# 
#      network:
#        plugin: calico
#        calico_network_provider:
#          cloud_provider: aws
# 
#   # To specify flannel interface
# 
#      network:
#        plugin: flannel
#        flannel_network_provider:
#          iface: eth1
# 
#   # To specify flannel interface for canal plugin
# 
#      network:
#        plugin: canal
#        canal_network_provider:
#          iface: eth1
# 
network: 
  options: 
    flannel_backend_type: "vxlan"
  plugin: "canal"
# 
#      services:
#        kube_api:
#          service_cluster_ip_range: 10.43.0.0/16
#        kube_controller:
#          cluster_cidr: 10.42.0.0/16
#          service_cluster_ip_range: 10.43.0.0/16
#        kubelet:
#          cluster_domain: cluster.local
#          cluster_dns_server: 10.43.0.10
# 
services: 
  etcd: 
    creation: "24h"
    extra_args: 
      election-timeout: "5000"
      heartbeat-interval: "500"
    retention: "72h"
    snapshot: true
  kube-api: 
    pod_security_policy: false
    service_node_port_range: "30000-32767"
  kubelet: 
    fail_swap_on: false
ssh_agent_auth: false
```

- Docker (use `docker info`):

```
Containers: 25
 Running: 15
 Paused: 0
 Stopped: 10
Images: 14
Server Version: 17.03.2-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-957.1.3.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.62 GiB
Name: <REDACTED>
ID: 7PN7:OVFQ:6X4K:N6O4:EMFQ:TU7G:DNGN:O4FG:KLR3:TSQX:GJN2:FUH6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://<REDACTED>:8080/
No Proxy: localhost,127.0.0.1,0.0.0.0,localaddress,10.0.0.0/8
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by dsargent3220** [comment]
We are experiencing this problem on one of our test clusters.  I just upgraded the Rancher that manages that cluster to 2.1.6 and upgraded the Kubernetes version from 1.11.5 to 1.11.6 and it is still doing it.  This is very frustrating as it makes this cluster unusable for just about any kind of workload.

**Comment by ArgonV** [comment]
I am seeing this too, with a 11 node RancherOS cluster (5 etcd/controlplane, 6 workers), and a vSphere cloud config. We are spanning two data centers in the same region. Everytime I loose a worker node, I have this:

```"For process kubelet, Env has changed from [RKE_CLOUD_CONFIG_CHECKSUM=........```

**Comment by ArgonV** [comment]
Does anyone know where this env var rke_cloud_config_checksum is set?

**Comment by alena1108** [comment]
According to @moelsayed, it should be resolved with https://github.com/rancher/rke/pull/1055

**Comment by moelsayed** [comment]
Can be validated in `2.1.7-rc3` and master.

**Comment by ArgonV** [comment]
@moelsayed<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_moelsayed&d=DwMCaQ&c=ODFT-G5SujMiGrKuoJJjVg&r=543gQwZQ37wtht6QPHrf7w&m=ApoQ3TjLSOgeOn1XarXaXXuYI2BG2qNUlHGiApw9KAE&s=noMPB2iVTkzwNAMggI4svVAzlYgLWqJBt2f3MQXAAwg&e=>, do you mean the bug exists in 2.1.7 rc3, or is resolved?

On Feb 26, 2019, at 1:58 PM, Mohamed El Sayed <notifications@github.com<mailto:notifications@github.com>> wrote:


Can be validated in 2.1.7-rc3 and master.

**Comment by moelsayed** [comment]
@ArgonV still exists in 2.1.7rc.

**Comment by sowmyav27** [comment]
validated in rancher:master (02/28/2019)
Able to launch an Azure cluster. All the nodes get to Active state. (Adding and deleting of worker nodes is working, and the nodes come to 'Active' state.

**Comment by ArgonV** [comment]
@sowmyav27, do you still see the logs for the cattle-node-agent, where the checksums are changing on the Azure worker nodes?

**Comment by sangeethah** [comment]
This  same issue was observed when testing Azure cluster with provider details set - https://github.com/rancher/rancher/issues/16893

**Comment by Oats87** [comment]
Hold on, I don't know if this issue is validated given your observations, those are not the symptoms that would cause it.

@sowmyav27 can you provide logs of the `cattle-node-agent` for validation?

**Comment by ArgonV** [comment]
I have found a stop-gap fix that I've tested on two vSphere node driver clusters, with node template provisioned VMs in vSphere, and using VMWare vSphere Volume storage class:

Add this to your cluster yaml, near the bottom
```
kubelet: 
    extra_args: 
      node-status-update-frequency: "10s"
    extra_env: 
      - "RKE_CLOUD_CONFIG_CHECKSUM=<your value>"
      - "RKE_CLOUD_CONFIG_CHECKSUM=<your value>"
```

You can get your checksum values by looking at your logs for the cattle node agents.

I have tested on clusters running Kubernetes v1.12.4 and v1.11.6

**Comment by Oats87** [comment]
@ArgonV Thanks for the note, did you need to specify the extra args twice for `RKE_CLOUD_CONFIG_CHECKSUM` like you have in your example?

**Comment by ArgonV** [comment]
@Oats87, yes I did. It didn't fix the issue with one RKE_CLOUD entry and comma separated values. It also didn't work with just one RKE_CLOUD entry and one env var. I forgot to mention that I also redeployed in the System project after I edited and updated the cluster:

cattle-cluster-agent
cattle-node-agent
nginx-ingress-controller

**Comment by galal-hussein** [comment]
Tested the issue with both 2.1.4 and master, can't see to reproduce the issue, here are my steps:

- Start Rancher (2.1.4 or master) using single node installation
- Start a vsphere cluster with vsphere cloud provider
- use custom cloud provider and edit yaml and add:
```
cloud_provider: 
  name: "vsphere"
  vsphereCloudProvider: 
    global: 
      datacenter: "xxxx"
      working-dir: "xxxx"
      insecure-flag: true
      password: "xxxxx"
      soap-roundtrip-count: 0
      user: "xxxxx"
      server: "xxxxx"
    workspace: 
      datacenter: "xxxx"
      default-datastore: "xxxxx"
      resourcepool-path: "xxxxxx"
      server: "xxxxx"
```

I can see cluster getting created with different roles including worker node.
But it seems the issue is resolved already with Azure https://github.com/rancher/rancher/issues/16893.

**Comment by alena1108** [comment]
@ArgonV 

> I have tested on clusters running Kubernetes v1.12.4 and v1.11.6

the fix for this issue is done for 2.2 version of rancher; it means it is available in master branch only, and not available in 2.1.x release. If you've done your testing on 2.1.6, could you give it a test on the latest master? `rancher/rancher:master` image.

**Comment by cjellick** [comment]
ive moved this back to test since our assertion is that this is fixed on master

**Comment by Oats87** [comment]
I ran the following scenario:

Rancher `v2.1.7` with a 3 node cluster, 1 controlplane+etcd and 2 worker. It was observed that the `Env has changed from` issue was occurring on my worker instances.

I upgraded to Rancher `master` and watched the agents for approximately 30 minutes, and did not see a single kubelet restart (bar from one where the agent cleaned up after a faulty reset). It appears that this issue has been resolved for `v2.2.0`
